File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        
        ptr1 = list1
        ptr2 = list2
        
        res = ListNode()
        dummy = res
         
        while ptr1 and ptr2:
            if ptr1.val < ptr2.val:
                res.next = ListNode(ptr1.val)
                ptr1 = ptr1.next
            else:
                res.next=ListNode(ptr2.val)
                ptr2 = ptr2.next
            res = res.next
            
        while ptr1:
            res.next = ListNode(ptr1.val)
            ptr1 = ptr1.next
            res = res.next
        
        while ptr2:
            res.next=ListNode(ptr2.val)
            ptr2 = ptr2.next
            res = res.next
        
        return dummy.next
    
            
        
                
                
                
        
        
        
        
        

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* head = nullptr, *cur = nullptr;
        if (list1 && list2) {
            if (list1->val > list2->val) {
                head = list2;
                list2 = list2->next;
            }
            else {
                head = list1;
                list1 = list1->next;
            }
            cur = head;

            while (list1 && list2) {
                if (list1->val > list2->val) {
                    cur->next = list2;
                    list2 = list2->next;
                }
                else {
                    cur->next = list1;
                    list1 = list1->next;
                }
                cur = cur->next;
            }

            while (list1) {
                cur->next = list1;
                list1 = list1->next;
                cur = cur->next;
            }

            while (list2) {
                cur->next = list2;
                list2 = list2->next;
                cur = cur->next;
            }

            return head;
        }
        else {
            if (list1) return list1;
            else if (list2) return list2;
            else return nullptr;
        }
    }
};